lib/src/rewrite.rs

```diff
@@ -158,12 +158,16 @@ pub fn rebase_commit_with_options(
             }
             EmptyBehaviour::AbandonAllEmpty => *parent.tree_id() == new_tree_id,
         };
-        if should_abandon {
-            mut_repo.record_abandoned_commit(old_commit.id().clone());
-            // Record old_commit as being succeeded by the parent for the purposes of
-            // the rebase.
+        // If the user runs `jj checkout foo`, then `jj rebase -s foo -d bar`, and we
+        // drop the checked out empty commit, then the user will unknowingly
+        // have done the equivalent of `jj edit foo` instead of `jj checkout
+        // foo`. Thus, we never allow dropping the working commit. See #2766 and
+        // #2760 for discussions.
```

Comment:
I don't think this is the right solution. I don't like to special-case the working copy like this. I think we should instead abandon the working copy and then create a new one on top, the same way we do if you just run `jj abandon`.

Reply:
That SGTM, I think I also prefer that slightly to what we did here. Though, I would guess that few people would ever notice the difference.

I mentioned the possibility of the following logic somewhere: if the change id of the working copy before is different from the change id after, do the equivalent of `jj new`.

This differs from this PR in what happens when the empty/emptied working copy has children and what happens to the description.

@matts1, WDYT?

Reply:
It's a good point that mimicking `jj abandon` here would probably make things feel more consistent to users (the ones who notice, consciously or not).

Reply:
The fix might be clearer once we've moved the updating of branches and working copies out of `rebase_next()` and into `rebase_all()` (or some helper called from there).

Reply:
I'm not sure how straightforward my suggestion is to implement, though, since the rebaser currently never creates commits out of nothing.

For now, the logic I suggested could go into the `rebase` CLI rather than the rebaser, but that can only last until rebasing with abandons is restricted to `jj rebase` only. I guess we could eventually teach `rebase_descendants_return_map` (or whatever that will become) that commits can be created out of thin air.

Reply:
> I'm not sure how straightforward my suggestion is to implement, though, since the rebaser currently never creates commits out of nothing.

Not sure what you mean, but we do create new working-copy commits here:
https://github.com/martinvonz/jj/blob/f169c99fb42bed7dba5bda7d5761e84b8308ecfa/lib/src/rewrite.rs#L512-L518

Reply:
I didn't know that, good point and good to know!

On second thought, what I said in my previous comment might indeed not be very relevant.

The point was that I don't think there's any way to represent this in the return value of `rebase_descendants_return_map` and similar places, but this only matters if we plan to rebase further commits on top of the working copy (which we don't).

Reply:
@martinvonz I don't think special-casing the working commit is avoidable. IIUC your proposal is to just ignore the working copy and unconditionally create a new empty commit when the rebase is finished? If so, that would fail when the working copy is non-empty, since the new working commit would be a new empty commit when the original working commit was non-empty

Reply:
No, my proposal is to make it behave *as if* you had rebased without `--skip-empty/-emptied` followed by abandoning the empty/emptied commits afterwards. I don't know how we should implement it :)

Reply:
Ah, I see. I'm not opposed to that, but seems like a lot of work for something that seems to be working as is, since I think it'd require a complete rewrite of the feature. 

If someone else wants to take it on, they're welcome to, but I'm happy with the current state of the feature, so I was just gonna leave as is.

Reply:
I can try my suggestion and see how hacky it turns out to be. I'm not sure when I'll get to it, we can file a bug so that we don't forget.

Reply:
This is now tracked as #2869 .